Plutarch/Api/V1/AssocMap.hs

```diff
@@ -480,7 +482,68 @@ punionWithData ::
     )
 punionWithData = phoistAcyclic $
   plam $ \combine x y ->
-    pcon $ PMap $ (pmatch (mapUnion # combine) \(MapUnionCarrier {merge}) -> merge) # pto x # pto y
+    pcon $ PMap $ (pmatch (mapUnion # combine) \(MapMergeCarrier {merge}) -> merge) # pto x # pto y
+
+{- | Intersect two 'PMap's applying the given function to any two values that
+ share the same key.
+-}
+pinsecWith ::
+  (POrd k, PIsData k, PIsData v) =>
+  Term s ((v :--> v :--> v) :--> PMap 'Sorted k v :--> PMap 'Sorted k v :--> PMap 'Sorted k v)
+pinsecWith = phoistAcyclic $
+  plam $
+    \combine -> pinsecWithData #$ plam $
+      \x y -> pdata (combine # pfromData x # pfromData y)
+
+mapInsecCarrier :: (POrd k, PIsData k) => Term s ((PAsData v :--> PAsData v :--> PAsData v) :--> MapMergeCarrier k v :--> MapMergeCarrier k v)
+mapInsecCarrier = phoistAcyclic $ plam \combine self ->
+  let mergeInsert = pmatch self \(MapMergeCarrier {mergeInsert}) -> mergeInsert
+      merge = pmatch self \(MapMergeCarrier {merge}) -> merge
+   in pcon $
+        MapMergeCarrier
+          { merge = plam $ \xs ys -> pmatch xs $ \case
+              PNil -> pnil -- diff to mapUnionCarrier here
+              PCons x xs' -> mergeInsert # x # xs' # ys
+          , mergeInsert = plam $ \x xs ys ->
+              pmatch ys $ \case
+                PNil -> pnil -- diff to mapUnionCarrier here
+                PCons y1 ys' ->
+                  plet y1 $ \y ->
+                    plet (pfstBuiltin # x) $ \xk ->
+                      plet (pfstBuiltin # y) $ \yk ->
+                        pif
+                          (xk #== yk)
+                          ( pcons
+                              # (ppairDataBuiltin # xk #$ combine # (psndBuiltin # x) # (psndBuiltin # y))
+                              #$ merge
+                                # xs
+                                # ys'
+                          )
+                          ( pif
+                              (pfromData xk #< pfromData yk)
+                              (mergeInsert # y # ys' # xs) -- diff to mapUnionCarrier here
+                              (mergeInsert # x # xs # ys') -- diff to mapUnionCarrier here
+                          )
+          }
+
+mapInsec :: forall k v s. (POrd k, PIsData k) => Term s ((PAsData v :--> PAsData v :--> PAsData v) :--> MapMergeCarrier k v)
+mapInsec = phoistAcyclic $ plam \combine -> punsafeCoerce pfix # (mapInsecCarrier # combine :: Term _ (MapMergeCarrier k v :--> MapMergeCarrier k v))
+
+{- | Intersect two 'PMap's applying the given function to any two data-encoded
```

Comment:
Instead of "applying", it could be more explicit: "merging values that share the same key, using the given function".